### PR TITLE
Add new blog post on structuring a large app

### PR DESCRIPTION
**Pull request by masenf**
https://dev.reflexcorp.dev/blog/2024-03-08-structuring-a-large-app/

☝️ link broken, maybe someone re-deployed dev from `main`?

### CODE REVIEW

File: blog/2024-03-08-structuring-a-large-app.md [hunk 1]
```diff
@@ -0,0 +1,391 @@
+---
+author: Masen Furer
+date: 2024-03-08
+title: "Structuring a Large App"
+description: "So your Reflex app is getting large? Here's some advice on how to lay it out."
+image: /blog/structure-large-app.jpg
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:6]: nit: we should resize the image, either here or in the blog.py code to make it a consistent size everywhere. Currently it's a bit large.

File: blog/2024-03-08-structuring-a-large-app.md [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+---
+
+# First Steps
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:9]: May be nicer has H2 ## since the top level title is an H1

File: blog/2024-03-08-structuring-a-large-app.md [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+Reflex makes it easy to get a simple web app running from a single python file:
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:11]: nit: Python

File: blog/2024-03-08-structuring-a-large-app.md [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    return rx.center(
+        rx.heading("Hello Reflex! 👋")
+    )
+
+
+app = rx.App()
+```
+
+However, as the app grows, keeping everything in a single module can quickly
+become messy and unmanageable. Because a Reflex app is just a python project,
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:28]: maybe move the second sentence to a new paragraph. I find shorter paragraphs are much more readable in these blogs. Also, capitalize Python everywhere

File: blog/2024-03-08-structuring-a-large-app.md [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+it can be organized into modules and packages (directory with `__init__.py`)
+and use the python import machinery you're likely already familiar with.
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:30]: style: machinery may be a scary word, maybe reword to "imported like any other python object"

File: blog/2024-03-08-structuring-a-large-app.md [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+# What Makes a Reflex App?
+
+When executing `reflex init` or `reflex run`, the framework will import **two** modules automatically.
+
+## rxconfig.py
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:36]: not sure if this is needed in this blog post, it may take focus away from the project structure (since this won't really be changed)

File: blog/2024-03-08-structuring-a-large-app.md [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+
+## \{app_name}/\{app_name}.py
+
+After loading the config, Reflex imports the main app module based on the
+`app_name`, which **must define a module-level global named `app` as an instance
+of `rx.App`**.  In the example above, this would be
+`example_big_app/example_big_app.py`.
+
+The main app module is responsible for importing all other modules that make up the
+app and defining `app = rx.App()`. **All other modules containing pages, state,
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:56]: The "see" portion is a bit confusing. Is this mostly needed because of the `rx.page` decorator? For example, other states may not necessarily have to be imported directly in the main app file, if it imports a page that imports the state downstream.

File: blog/2024-03-08-structuring-a-large-app.md [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+and models MUST be imported by the main app module or package** for Reflex to be able to
+"see" these parts of the app.
+
+# How to Split Up the App
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:60]: Maybe we can start with the default app structure, and incrementally build it up. Like a story. For example, let's say we want to add multiple pages, we introduces the `pages/` folder. Then we want reusable components, we introduce the `components` folder. Then at the end we can show the final structure.

The way it is now, it kind of goes from the solution backwards, but I feel motivating it might be nice and then the user can digest it incrementally.

File: blog/2024-03-08-structuring-a-large-app.md [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+│  ├─ example_big_app.py
+│  ├─ models.py
+│  ├─ state.py
+│  ├─ template.py
+├─ uploaded_files/
+├─ requirements.txt
+├─ rxconfig.py
+```
+
+## example_big_app/models.py
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:92]: I would start with the pages (then components, then models) maybe. Just to go from most used feature to least used feature.

File: blog/2024-03-08-structuring-a-large-app.md [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+
+I prefer to implement all database models in a single file to make it easier to
+define relationships and understand the entire schema. However, if the schema is
+very large, it might make sense to have a `models` package with individual
+models defined in their own modules. At any rate, defining the models separately
+allows any page or component to import and use them without circular imports.
+
+## example_big_app/state.py
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:100]: We don't recommend substates anymore, so I think we can remove this. And just say this is where you can put any global state.

File: blog/2024-03-08-structuring-a-large-app.md [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+
+## example_big_app/components
+
+This package contains reusable parts of the app, for example headers, footers,
+and menus. If a particular component requires state, the substate may be defined
+in the same module for locality. Any substate defined in a component module
+should only contain fields and event handlers pertaining to that individual
+component. 
+
+## example_big_app/template.py
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:113]: This is great, and deserves its own section in the main docs

File: blog/2024-03-08-structuring-a-large-app.md [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+
+## example_big_app/\__init\__.py
+
+The `__init__.py` file defines the top-level package. This is a great place to import all state, models, and
+pages that should be part of the app. Typically, components and helpers
+do not need to imported, because they will be imported by pages that use them (or they would be unused).
+
+```python
+from . import state, models
+from .pages import index, login, post, product, profile, schedule
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:194]: Are we recommending the decorator primarily now?

File: blog/2024-03-08-structuring-a-large-app.md [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+            rx.button("< Prev", on_click=PostsState.go_to_previous),
+            rx.button("Next >", on_click=PostsState.go_to_next),
+            justify="between",
+        ),
+        rx.moment(interval=PostsState.refresh_tick, on_change=PostsState.tick, display="none"),
+        width="100%",
+    )
+```
+
+# Component Reusability
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:331]: This should be in the `components/` folder section

File: blog/2024-03-08-structuring-a-large-app.md [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+
+The primary mechanism for reusing components in Reflex is to define a function that returns
+the component, then simply call it where that functionality is needed.
+
+Component functions typically should not take any State classes as arguments, but prefer
+to import the needed state and access the vars on the class directly.
+
+## functools.lru_cache
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:339]: I think we can have this in the docs, but not needed as part of "project structure"

File: blog/2024-03-08-structuring-a-large-app.md [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+        *[
+            foo()
+            for _ in range(100)
+        ],
+        direction="row",
+        wrap="wrap",
+    )
+```
+
+## External Components
```
Comment (picklelo) [inline blog/2024-03-08-structuring-a-large-app.md:374]: We can remove this too, and just focus on their core project structure

File: blog/2024-03-08-structuring-a-large-app.md (hunk 12 above)
Comment (masenf) [reply blog/2024-03-08-structuring-a-large-app.md:194]: decorator is really nice for delegating control over page title, on loads, etc to the page module without have to deal with circular imports. i would say it's recommended

File: blog/2024-03-08-structuring-a-large-app.md (hunk 14 above)
Comment (masenf) [reply blog/2024-03-08-structuring-a-large-app.md:339]: removed from post

File: blog/2024-03-08-structuring-a-large-app.md (hunk 10 above)
Comment (masenf) [reply blog/2024-03-08-structuring-a-large-app.md:100]: i definitely think substates still have their place, for example when you need to access most of the vars and event handlers of another state.

base states that have top-level app vars are also important and i dont think they're going away, even though we now have a different mechanism for state access.

File: blog/2024-03-08-structuring-a-large-app.md (hunk 6 above)
Comment (masenf) [reply blog/2024-03-08-structuring-a-large-app.md:36]: personally i think it helps set the stage for how reflex "ingests" a project. rxconfig.py and the main app module are really important pieces, everything else builds on top of that

File: blog/2024-03-08-structuring-a-large-app.md (hunk 1 above)
Comment (masenf) [reply blog/2024-03-08-structuring-a-large-app.md:6]: used alek's image

**Comment by masenf** [comment]
@picklelo updated according to comments

File: blog/2024-03-15-structuring-a-large-app.md [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+
+## Breaking the App into Modules
+
+The key to good organization as the app grows is to break down all parts of the app into
+smaller, more manageable pieces, and organize them into logical packages that avoid circular
+dependencies.
+
+### Pages
+
+#### example_big_app/template.py
```
Comment (picklelo) [inline blog/2024-03-15-structuring-a-large-app.md:71]: Maybe before introducing templates, show factoring out pages into multiple files. Then we can note that the pages have something in common which leads to the template. (pretty much reordering this with the next section, since multiple pages is the more basic use case than templating)

**Review by picklelo — APPROVED** [review]
looks great! one suggestion, but I'm good to merge and publish this week

**Review by picklelo — APPROVED** [review]
